wannierberri/__system.py

```diff
@@ -409,6 +409,66 @@ def check_hermitian(self,XX):
         else:
             print (f"{XX} is missing,nothing to check")
 
+    def set_structure(self, positions, atom_numbers, magnetic_moments=None):
+        """
+        Set atomic structure of the system.
+
+        Parameters
+        ----------
+        positions : (num_atom, 3) array_like of float
+            Atomic positions in fractional coordinates.
+        atom_numbers: (num_atom,) list of int
+            Integer numbers to distinguish species.
+        magnetic_moments: (num_atom, 3) array_like of float (optional)
+            Magnetic moment vector of each atom.
+        """
+        if len(positions) != len(atom_numbers):
+            raise ValueError("length of positions and atom_numbers must be the same")
+        if magnetic_moments is not None and len(positions) != len(magnetic_moments):
+            raise ValueError("length of positions and magnetic_moments must be the same")
+        self.positions = positions
+        self.atom_numbers = atom_numbers
+        self.magnetic_moments = magnetic_moments
+
+    def get_spglib_cell(self):
+        """Returns the atomic structure as a cell tuple in spglib format"""
+        try:
+            if self.magnetic_moments is None:
+                return (self.real_lattice, self.positions, self.atom_numbers)
+            else:
+                return (self.real_lattice, self.positions, self.atom_numbers, self.magnetic_moments)
+        except AttributeError:
+            raise AttributeError("set_structure must be called before get_spglib_cell")
+
+    def set_symmetry_from_structure(self):
+        """
+        Set the symmetry group of the :class:`System`. Requires spglib to be installed.
+        :meth:`System.set_structure` must be called in advance.
+
+        For magnetic systems, symmetries involving time reversal are not detected because
+        spglib does not support time reversal symmetry for noncollinear systems.
+        """
+        import spglib
+
+        spglib_symmetry = spglib.get_symmetry(self.get_spglib_cell())
+        symmetry_gen = []
+        for isym in range(spglib_symmetry["rotations"].shape[0]):
+            # spglib gives real-space rotations in reduced coordinates. Here,
+            # 1) convert to Cartesian coordinates, and
+            # 2) take transpose to go to reciprocal space.
+            W = spglib_symmetry["rotations"][isym]
+            Wcart = self.real_lattice.T @ W @ np.linalg.inv(self.real_lattice).T
+            R = Wcart.T
+            symmetry_gen.append(Symmetry(R))
+
+        if self.magnetic_moments is None:
+            symmetry_gen.append(TimeReversal)
+        else:
+            print("Warning: spglib does not support time reversal symmetry for noncollinear "
```

Comment:
@jaemolihm , did you try testing `spglib` mith magnetic moments ? does it work?

I tried and it looks to neglect them, e.g. 

```
In [3]: spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]],[ [-0.1,-0.1,-0.1],[0.1,0.1,0.1]],[1,1],[-1,1]))                                                                                                                                                                                         
Out[3]: 
{'rotations': array([[[ 1,  0,  0],
         [ 0,  1,  0],
         [ 0,  0,  1]],
 
        [[-1,  0,  0],
         [ 0, -1,  0],
         [ 0,  0, -1]]], dtype=int32),
 'translations': array([[0., 0., 0.],
        [0., 0., 0.]]),
 'equivalent_atoms': array([0, 0], dtype=int32),
 'primitive_lattice': array([[1., 0., 0.],
        [0., 2., 0.],
        [0., 0., 3.]])}

In [4]: spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]],[ [-0.1,-0.1,-0.1],[0.1,0.1,0.1]],[1,1],[1,1]))                                                                                                                                                                                          
Out[4]: 
{'rotations': array([[[ 1,  0,  0],
         [ 0,  1,  0],
         [ 0,  0,  1]],
 
        [[-1,  0,  0],
         [ 0, -1,  0],
         [ 0,  0, -1]]], dtype=int32),
 'translations': array([[0., 0., 0.],
        [0., 0., 0.]]),
 'equivalent_atoms': array([0, 0], dtype=int32),
 'primitive_lattice': array([[1., 0., 0.],
        [0., 2., 0.],
        [0., 0., 3.]])}

```

Result is unaffected by the fourth entry of `cell`

Reply:
You are right. `spglib` does not work with magnetic moments.

Reply:
then, maybe, for this PR just disable magnetic moments, and write in the docs that only TRS systems are allowed. implement magnetic systems later.

Reply:
Sorry, I misunderstood what you meant.

I think your are correct that spglib does not work for collinear magnetic moments as input. (It says it works, so maybe I am using wrong input.)

Spglib does work for noncollinear magnetic moments as input. For collinear systems, one can set the magnetic moments to e.g. [0., 0., 1.]. For example (based on your code, but changed only the magnetic moments argument to a vector:
```python3
>>> spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]], [ [-0.1,-0.1,-0.1], [0.1,0.1,0.1]], [1,1], [[0.,0.,1.], [0.,0.,1.]]))
{'rotations': array([[[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]]], dtype=int32), 'translations': array([[0., 0., 0.],
       [0., 0., 0.]]), 'equivalent_atoms': array([0, 0], dtype=int32), 'primitive_lattice': array([[1., 0., 0.],
       [0., 2., 0.],
       [0., 0., 3.]])}
>>> spglib.get_symmetry(([[1,0,0],[0,2,0],[0,0,3]], [ [-0.1,-0.1,-0.1], [0.1,0.1,0.1]], [1,1], [[0.,0.,1.], [0.,0.,-1.]]))
{'rotations': array([[[1, 0, 0],
        [0, 1, 0],
        [0, 0, 1]]], dtype=int32), 'translations': array([[0., 0., 0.]]), 'equivalent_atoms': array([0, 1], dtype=int32), 'primitive_lattice': array([[1., 0., 0.],
       [0., 2., 0.],
       [0., 0., 3.]])}
```

The limitation in this case is that spglib misses out the symmetry operations that involve time reversal (such as  time reversal * C2z). In short, spglib works but finds only a subset of the actual symmetry of the system.

The case with magnetic moments is actually being tested for Fe: see test_symmetry_spglib_Fe.

What I can do is to add an error if `magnetic_moments` is set to a number, not a vector. I will do this in this PR soon.

Reply:
> What I can do is to raise an error if magnetic_moments is set to a list of numbers, not a list of vectors.

I think it is a way to go, not to complicate things at this point

> Or, one can convert from numbers to vectors internally

 in that case one needs to specify the quantization axis (which is not always `z`, but may have a default value `[0,0,1]`

>  The problem is that spglib does not tell you whether the given symmetry operation involves time reversal or not

Then it is (almost) useless information for us. Well, for the TR-even tensors it may be used. But I suggest to consider only the non-collinear case. Anyway, in a collinear case the spin-up and spin-down subsystems may be considered independently. 

So, just make sure magnetic moments passed to spglib are vectors not numbers, either way.